### DISCUSSION TITLE
Mount machine.SDCard to existing SPI bus

### DISCUSSION THREAD

**Discussion by SL2021-Dev** [micropython/micropython#12744] [ESP32]
I am trying to use machine.SDCard with lv_micropython ili9XXX.py and xpt2046.py drivers

If I mount the SDCard first and do a print(os.listdir('/sd/data')) so far so good
then initialize the ili9341 then run the print(os.listdir('/sd/data')) again...
I get an...  E (38150) sdmmc_cmd: sdmmc_read_sectors_dma: sdmmc_send_cmd returned 0x107
Why does the SDCard fail??
**Example 1**
```
import lvgl as lv
from ili9XXX import ili9341
from xpt2046 import xpt2046
from machine import SDCard, SPI, Pin
import os

try:
	os.mount(SDCard(slot=2, width=1, miso=19, mosi=23, sck=18, cs=14, freq=20_000_000), '/sd')
	print(f"/sd/data: {os.listdir('/sd/data')}")
except Exception as e:
	print(f'SD Error\n{e}')

disp = ili9341(spihost=2, miso=-1, mosi=-1, clk=-1, cs=15, mhz=20, dc=13, rst=-1, power=-1, factor=48, half_duplex=False, double_buffer=False, backlight=27, backlight_on=1)
```
**output...**
>>>
>>>
/sd/data: ['test.txt']
Single buffer
ILI9341 initialization completed
Enable backlight
♦♦>
MicroPython v1.20.0-702-g1ed8b7c0d-dirty on 2023-10-18; ESP32 module (spiram) with ESP32
Type "help()" for more information.
>>>
>>> print(f"/sd/data: {os.listdir('/sd/data')}")
E (38150) sdmmc_cmd: sdmmc_read_sectors_dma: sdmmc_send_cmd returned 0x107
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 2] ENOENT
>>> 

What I want to to do is reverse the order of conecting devices to the SPI bus but I need a method to attach
**machine.SDCard to an existing SPI bus**
**Example 2**
```
import lvgl as lv
from ili9XXX import ili9341
from xpt2046 import xpt2046
from machine import SDCard, SPI, Pin
import os

disp = ili9341(spihost=2, miso=19, mosi=23, clk=18, cs=15, mhz=20, dc=13, power=-1, rst=-1, factor=48, half_duplex=False, double_buffer=False,backlight=27, backlight_on=1)

os.mount(SDCard(slot=2, miso=None, mosi=None, sck=None, cs=14, freq=20_000_000), '/sd')
print(f"/sd/data: {os.listdir('/sd/data')}")
```
**output...**
>>>
>>>
Single buffer
ILI9341 initialization completed
Enable backlight
♦Traceback (most recent call last):
  File "<stdin>", line 49, in <module>
OSError: (-259, 'ESP_ERR_INVALID_STATE')
♦>
MicroPython v1.20.0-702-g1ed8b7c0d-dirty on 2023-10-18; ESP32 module (spiram) with ESP32
Type "help()" for more information.
>>>
>>>

Is there a method ????
Or will I get the same output.

**Comment by jimmo** [comment]
@SL2021-Dev Please see [Welcome (Please read before posting)](https://github.com/orgs/micropython/discussions/9111) for how to format code in posts.

> I need a method to attach machine.SDCard to an existing SPI bus

You can do this by using the Python sdcard driver instead of `machine.SDCard`. This works with `machine.SPI` instances. See https://github.com/micropython/micropython-lib/tree/master/micropython/drivers/storage/sdcard

**Reply by jimmo** [reply]
You might also want to reach out to the author of the ili9341 driver...

**Comment by SL2021-Dev** [comment]
The author quit !

[@amirgon](https://forum.lvgl.io/u/amirgon) just let me know that he can’t continue the maintenance of the the MicroPython biding due to other engagements.
So we are looking for a new maintainer for the MicroPython binding.

So I took your advice and used the sdcard.py, all three devices attach to the SPI bus !!!
But when I list... /sd  or /sd/data I get INTERMITTANT Errors...   File "sdcard.py", line 256, in readblocks OSError: [Errno 5] EIO

I believe there's some kind of collision between the sd card and the display or touch driver.
Do you have any thoughts on how I can correct this issue.

links to Disp and Touch drivers
https://github.com/lvgl/lv_binding_micropython/blob/6369315bb7bc3dc47f0b2cc33f7e7f8d830a5749/driver/esp32/ili9XXX.py
https://github.com/lvgl/lv_binding_micropython/blob/6369315bb7bc3dc47f0b2cc33f7e7f8d830a5749/driver/esp32/xpt2046.py

```
import lvgl as lv
from ili9XXX import ili9341
from xpt2046 import xpt2046
import os, time, sdcard
from machine import SPI, Pin

sd = sdcard.SDCard(SPI(2), Pin(14), baudrate=10_000_000)
os.mount(sd, '/sd')

lv.init()
disp = ili9341(spihost=2, miso=-1, mosi=-1, clk=-1, cs=15, mhz=10, dc=13, rst=-1, power=-1, factor=48, half_duplex=False, double_buffer=True, backlight=27, backlight_on=1)
touch = xpt2046(spihost=2, cs=5, mhz=10, half_duplex=False)

print(os.listdir('/sd'))
print(os.listdir('sd/data'))


'''
OUTPUT......
>>>
>>>
Double buffer
ILI9341 initialization completed
Enable backlight
['System Volume Information', 'data']
['test.txt']
Ready
♦♦>
MicroPython v1.20.0-702-g1ed8b7c0d-dirty on 2023-10-18; ESP32 module (spiram) with ESP32
Type "help()" for more information.
>>>
>>> print(os.listdir('sd/data'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sdcard.py", line 256, in readblocks
OSError: [Errno 5] EIO
>>>

'''
```

**Comment by SL2021-Dev** [comment]
So digging even deeper I found out that if I comment out the touch driver that listing the directories work.
So this is narrowed down now to the touch driver as being the culprit
